Feature/1632 label underline strike

alanag13 opened the pull request:
### Description of Change ###

Closes #1632.
Adds support for a Bindable `TextDecorations` property (Underline/Strikethrough) for Labels and Spans. 

### API Changes ###

Added:
 - enum TextDecorations `{ None, Underline, StrikeThrough }`
 - TextDecorations Label.TextDecorations { get; set; } // Bindable Property

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

kicsiede committed:
BUGFIX: leaking reference to popped Page in PopAsync()

fixed by using a private wait handle the same way as when there is a pending task

kicsiede committed:
Update NavigationUnitTest.cs

alanag13 committed:
Android implementation

alanag13 committed:
add UWP + WPF

alanag13 committed:
Fix WPF

alanag13 committed:
fix for backwards compatibility

alanag13 committed:
iOS and (probably) macOS

alanag13 committed:
Tizen

alanag13 committed:
remove unused prop

alanag13 committed:
refactor iOS/macOS

adamped commented on Xamarin.Forms.Core/Label.cs:
```diff
@@ -32,6 +32,8 @@ public class Label : View, IFontElement, ITextElement, ITextAlignmentElement, IE
 
 		public static readonly BindableProperty FontAttributesProperty = FontElement.FontAttributesProperty;
 
+		public static readonly BindableProperty TextDecorationsProperty = BindableProperty.Create("TextDecorations", typeof(TextDecorations), typeof(Label), TextDecorations.None);
```
You can actually use nameof(TextDecorations) now. They just didn't use it before, because XF started before the feature was added to C#

alanag13 replied:
sure, I'm aware of the feature but was trying to copy existing code styles. Thanks!

adamped replied:
What I was trying to say, is the XF team are happy for you to start using the nameof feature now :) They are hoping to update all in the future.

adamped commented on Xamarin.Forms.Platform.UAP/LabelRenderer.cs:
```diff
@@ -152,6 +157,34 @@ protected override void OnElementPropertyChanged(object sender, PropertyChangedE
+			{
+				if (_underlineSet)
+				{
+					Control.TextDecorations = Control.TextDecorations & ~Windows.UI.Text.TextDecorations.Underline;
```
This part has me a little concerned. Is this basically saying that whenever UpdateTextDecorations is called, it just flips both underline and strikethrough at the same time, back and forth? Shouldn't you need to check if the TextDecoration has actually been removed first?

alanag13 replied:
_underlineSet and _strikeThroughset are for backwards compatibility, I don't want to remove a strikethough or underline unless someone has explicitly set it using the new BP.

Youre right that what I'm doing is clearing both flags every time, but then I'm adding back the ones that are present on the BP -- so if one flag is removed it won't be added back.

alanag13 replied:
Unless I'm misunderstanding something `& ~` doesn't flip the bit, it sets it to 0 if it is 1. (`^=` flips the bit).

alanag13 replied:
I do realize now though that it would be more concise to do `Control.TextDecorations &= ~Flag`, so I'll go ahead and change that for now.

alanag13 replied:
okay I must have just been looking at this code for too long -- I understand your comments now and will make it so that these are only removed if they aren't on the BP, not sure how I missed this.

adamped commented on Xamarin.Forms.Platform.WPF/Renderers/LabelRenderer.cs:
```diff
@@ -53,6 +58,40 @@ protected override void UpdateBackground()
+			{
+				var newTextDecorations = new System.Windows.TextDecorationCollection(Control.TextDecorations);
+
+				if (_underlineSet)
```
Same again.

adamped commented on Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs:
```diff
@@ -156,6 +163,53 @@ protected override void OnElementPropertyChanged(object sender, PropertyChangedE
+#endif
+				var range = new NSRange(0, newAttributedText.Length);
+
+				if (_strikeSet)
```
Just because I haven't delved into this before, do you have to wipe everything that add it again, every time?

alanag13 replied:
On some platforms there wasn't an easy way to check if a particular style was set, so you would have to remove them all and then add back the ones you want. I would think that 99% of the time we wouldn't be combining Underline and strike together so this shouldn't cause unnecessary overhead of removing stuff that we don't need to.

Technically for the enum-based ones I could take the approach of only removing them if they are not in the BP, but then I would also presumbaly have to add logic to not add them in this case. That seemed a little over the top to me, but I can switch those if you want. For iOS and WPF I don't think there's really a good way to not do this though.

alanag13 commented:
GTK is still left and I still need to do `Span` but I'm waiting on #1850 for that.
CSS also still remaining.

I went down the path that was originally suggested in the issue at first, but I ended up deviating from the original implementation suggestion of modifying the `FontAttributes` enum for a couple reasons:

 - `Underline` and `StrikeThrough` are not valid css terms for `font-style`, which is what `FontAttributes` is currently bound to.
- Likewise, `Bold` and `Italic` are not valid css terms for `text-decoration`, which is what we want people to be able to use for this 
- `IFontElement` applies to loads of stuff that is not Spans/Labels where in some cases would be not possible and other cases impractical to implement. I'm thinking `Button` might be okay to add this to and would do so if you agree, but I don't see much of a case for any of the other `IFontElement` controls.
- As evidenced by many APIs and css itself, text decorations are not considered part of a font (they do not change the physical appearance of any character), but rather an effect applied to text. If you want to combine a font style and a text decoration, I think it is more discoverable / logical to do:
````C#
label.FontAttributes = FontAttributes.Bold;
label.TextDecorations = TextDecorations.Underline;
````
rather than:
````C#
label.FontAttributes = FontAttributes.Bold | FontAttributes.Underline;
````

adamped commented:
But otherwise I am very excited by this update. The most requested feature in my poll. I've wanted it numerous times. Now if I can just get that Tappable Span through, displaying a link inside a label, will finally be able to happen. So many people ask how to do this, and it's been a common pain point for years.

adamped commented:
I agree on your comments. Splitting them into TextDecorations, seems more logical, and how other platforms manage it.

alanag13 committed:
use nameof on BP

alanag13 committed:
only remove native text decorations if BP show it is not set (instead of removing every time and adding back desired ones)

alanag13 committed:
bug fixes and add binding gallery demo

alanag13 commented:
@adamped I made some changes that I think capture your concerns and also added some "tap to toggle" demos to the legacy label screen of the gallery.

jassmith commented on Xamarin.Forms.Platform.Tizen/Renderers/LabelRenderer.cs:
```diff
@@ -63,6 +64,13 @@ Native.FormattedString ConvertFormattedText(FormattedString formattedString)
 
+		void UpdateTextDecorations()
+		{
+			Control.Strikethrough = (Element.TextDecorations & TextDecorations.StrikeThrough) != 0;
```
Double GetValue call here

jassmith commented on Xamarin.Forms.Core/FontAttributes.cs:
```diff
@@ -9,4 +9,12 @@ public enum FontAttributes
+	{
+		None = 0,
+		Underline = 1 << 0,
+		StrikeThrough = 1 << 1
```
No capital T please

jassmith replied:
Switch underline and strikethrough to match UWP enum values

jassmith commented on Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs:
```diff
@@ -156,6 +163,53 @@ protected override void OnElementPropertyChanged(object sender, PropertyChangedE
 
+		void UpdateTextDecorations()
+		{
+			var textDecorations = Element.TextDecorations;
```
Why do you fetch this before checking IsSet?

jassmith commented on Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs:
```diff
@@ -156,6 +163,53 @@ protected override void OnElementPropertyChanged(object sender, PropertyChangedE
+		void UpdateTextDecorations()
+		{
+			var textDecorations = Element.TextDecorations;
+			if (Element.IsSet(Label.TextDecorationsProperty))
```
If (!Element.IsSet ....)
  return;

StephaneDelcroix committed:
[XamlC] Find the real VisualState parent (#2036)

`VisualStateGroupList` can be implicit, and it looks like tht wasn't
tested. This adds a test, and a fix.

 - fixes #2034

StephaneDelcroix committed:
[C] do not return IsSet for BP with defaultVC (#2055)

BindableProperties with a defaultValueCreator are never in the default
(as defined by the defaultValue parameter) state. IsSet now reflects
that.

To allow styling properties with defaultVC, this adds a new flag in
BindableContextAttributes.

StephaneDelcroix committed:
[C] Update Bindings with "BindingContext" in path (#2053)

* [C] Update Bindings with "BindingContext" in path

If the BindingContext is set using SetInheritedBindingContext, INPC
isn't fired, and bindings containing the "BindingContext" string in
the path ends up not being notified of the Change.

In addition to listening to PropertyChanged, this also hooks into
BindingContextChanged, and update the bindings accordingly.

 - fixes #2019

* back to last year

hartez committed:
Disable AndroidUseLatestPlatformSdk so VS stops auto-updating Android projects;
Remove XF.targets imports from projects which don't need it;
Make XF.targets imports conditional on existence of XFBT DLL to avoid errors

StephaneDelcroix committed:
Merge branch '15-5' into 3.0.0

StephaneDelcroix committed:
Merge branch '3.0.0'

rookiejava committed:
Support to change color on Slider for tizen backend (#2072)

hartez committed:
Revert "Disable AndroidUseLatestPlatformSdk so VS stops auto-updating Android projects;"

This reverts commit 3279709d950d0ea178f3ee812d28c2a91fd5d9d4.

Commit:
Merge branch '3.0.0'

rmarinho committed:
[Build] Update submodule to 15.5.7 (#2079)

rmarinho committed:
Merge branch '3.0.0'

BradChase2011 committed:
[macOS] Alternate picker.  Changed native control to NSPopupButton. (#1831)

* [macOS] Alternate picker.  Changed native control to NSPopupButton.

* *Fixed potential memory leak.

StephaneDelcroix committed:
[Xaml] better error reporting on x:Reference (#2089)

- fixes #2007

StephaneDelcroix committed:
[XamlC] report undeclared xmlns on custom markups (#2076)

- fixes #2063

StephaneDelcroix committed:
[XamlC] report missing ctor param (#2078)

for parameterized ctors, like Style.

 - fixes #2064

hartez committed:
Fix VS errors when loading clean solution; prevent VS rewriting Android project files (#2083)

* Disable AndroidUseLatestPlatformSdk in VS so VS stops auto-updating Android projects;
Remove XF.targets imports from projects which don't need it;
Make XF.targets imports conditional on existence of XFBT DLL in VS to avoid errors

* More consistent check for VS

* Apply nicer VS check logic to Xamarin.Forms.Xaml.UnitTests.csproj

* Fix missing "'"

* Re-add XF.targets imports to PagesGallery native projects

hartez committed:
[UWP] Thumb color and image for Slider (#2065) fixes #1684

* Add demo/test pages for new properties

* Test attributes for new Slider properties

* Implement slider thumb color and thumb image on UWP

* Use nameof for ThumbImage property

Commit:
Mark FormsEditText ctors public (#2103)

PureWeen committed:
Editor Auto Resize with Text Changes (#2091)

* Editor Auto Resize with Text Changes 1733

* comment changes

rmarinho committed:
[UWP] MDP should respect HasNavigationBar for current page (#2101) fixes #1437

* [UWP] MDP renderer should respect HasNavigationPage

* Update MasterDetailControl.cs

Commit:
[Android] Fix issue where compressed layouts couldn't be removed from the visual hierarchy (#2066) Fixes #1601

* Fix issue where compressed layouts couldn't be removed from the visual hierarchy on Android

* Dispose child packagers

* Hopefully fix 1601 UITest

* Remove invalid check

rmarinho committed:
[Enhancement] Entry ReturnType (#2029)

* [Core] Add ReturnType to Entry

* [Controls] Add Entry ReturnType gallery

* [Android] Entry ReturnType implementation

* [iOS] Entry ReturnType implementation

* [Docs] Docs for ReturnType

* [UWP] Entry ReturnType implementation

* [Core] Make sure we respect command CanExecute fixes #1697

* [Docs] Fix docs

* [Tizen] Entry ReturnType implementation

* [Core] Add missing defaults to BP on entry

* [Tests]Add unit test to entry Returntype command

* [iOS,Android,Tizen,UWP] Better Exception for ReturnType

* [Android] Handle completed event with new return type

* [ControlGallery]Fix ReturnType example

* [UnitTests] Remove extra Test attribute

rmarinho committed:
[Nuget] Update nuspec's with msft info (#2108)

rmarinho committed:
Merge branch '3.0.0'

ntherning committed:
Added the Entry.IsTextPredictionEnabled property (#2038) fixes #1677 

Makes it easier for users to disable text prediction/autocomplete/autosuggest.
Fixes #1677.

PureWeen committed:
The Dx for the shadow isn't being accounted for with the PaddingLeft of the background drawable

rmarinho committed:
[iOS] Set UITabBarItem colours when they are added  (#2109)

* [Controls] Add reproduction for #1323

* [iOS] Update UITabBar and text color after children pages changed #fixes 1323

rmarinho committed:
[iOS] Use UINavigationBar.Appearance.LargeTitleTextAttributes if specified (#2095)

* [Controls] Set UINavigationBar LargeTitleText appearance

* [iOS] Don't copy title attributes if user specified appearance for LargeTitles
fixes #1804

* [Controls] Only use LargeTitles if on iOS11

rmarinho committed:
Merge branch '3.0.0'

Commit:
Try/Catch assembly loading so previewer doesn't have a fit (#2106)

* Try/Catch assembly loading so previewer doesn't have a fit

* add warnings for failed dll loads

rmarinho committed:
[iOS] Encode url for UIWebView so it doesn't fail with non-ascii characters  (#2111)

* [Controls] Add reproduction for issue #1583

* [iOS] Encode url so it doesn't fail with non-ascii characters fixes #1583

* Update Issue1583_1.cs

myroot committed:
Enable KeyboardFlags and IsSpellCheckEnabled in Tizen renderer

rmarinho committed:
Merge pull request #2092 from xamarin/fix-bugzilla-31171 fixes #1429

[Core] Fixed leaking reference to popped Page in PopAsync()

PureWeen committed:
Merge pull request #2114 from xamarin/ShadowDxButtonDrawable

[Android] Button PaddingLeft now accounts for ShadowDx

StephaneDelcroix committed:
Merge branch '15-5' into 3.0.0

StephaneDelcroix committed:
Merge branch '3.0.0'

rmarinho committed:
Merge pull request #2116 from myroot/tizen-auto-capital

[Tizen] Enhance the Entry renderer of tizen backend

hartez committed:
Update submodule

alanag13 committed:
pr feedback
